text/0016-standard-modes.md

```diff
@@ -0,0 +1,292 @@
+
+More specific flight behaviour/modes are defined in _custom modes_, which are specific to each flight stack.
+Mechanisms are provided to set the base and custom modes ([MAV_CMD_DO_SET_MODE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_MODE)).
+The `HEARTBEAT` contains the current/active `base_mode` and `custom_mode`.
+There are no mechanisms query the available custom modes.
+
+Most autopilots implement a set of custom flight modes that have very similar behaviour or intent, for example: position-hold mode, altitude-hold mode, mission mode, safety return/RTL.
+
+However because these are all identified by different custom mode identifiers on different flight stacks, there is no way to be sure what these "mean" without pre-existing knowledge.
+
+
+# Detailed Design
+
+The design proposes a small set of modes that are (broadly speaking) commonly implemented across flight stacks, along with mechanisms to:
+- determine what modes are available
+- set the standard mode
+- infer the current standard mode, if active.
+
+
+## Standard modes
+
+Standard modes are those that are implemented in a similar way by most flight stacks, and which a GCS can present as more or less the same thing for any flight stack.
+
+The precise mechanics and behaviour do not have to be identical, but the broad intent has to be similar, and any differing behaviour across flight statcks should not be unexpected by a user.
+
+A flight stack is not _required_ to support all of these modes.
+
+The standard modes will be defined in an enum `MAV_STANDARD_MODE`, which has an initial value `0` is an enum that indicates "not a standard mode".
+
+The proposed initial set of modes is:
+```xml
+    <enum name="MAV_STANDARD_MODE">
+      <description>Standard modes with a well understood meaning across flight stacks and vehicle types.
+        For example, most flight stack have the concept of a "return" or "RTL" mode that takes a vehicle to safety, even though the precise mechanics of this mode may differ.
+        Modes may be set using MAV_CMD_DO_SET_STANDARD_MODE.
+      </description>
+      <entry value="0" name="MAV_STANDARD_MODE_NON_STANDARD">
+        <description>Non standard mode.
+          This may be used when reporting the mode if the current flight mode is not a standard mode.
+        </description>
+      </entry>
+      <entry value="1" name="MAV_STANDARD_MODE_POSITION_HOLD">
+        <description>Position mode (manual).
+          Position-controlled and stabilized manual mode.
+          When sticks are released vehicles return to their level-flight orientation.
+          Hovering vehicles actively brake and hold both position and altitude against wind and external forces,
+          Forward-traveling vehicles maintain current track and altitude.
+        </description>
```

Comment:
This needs re-work.  Hovering vs Forward-travelling isn't really clear, and whilst essentially c&p from PX4, the forward-travelling description isn't intuitive for fixed wing aircraft.  "Position Hold" in fixed wing is more intuitively circling a fixed point such as the point the mode was selected, and for rovers just maintaining current position.
It might be better to limit the first cut of "standard modes" to just multicopter, to avoid unwieldy/compromised descriptions.

Reply:
I have limited to MC, FW and VTOL.

Reply:
@auturgy In theory I think it should be possible to define what this mode means for other vehicle types, and we should have a go at it to see where the pain points are.

How about (for the hold part of the mode)

> Boats (USVs) maintain the current track against wind, waves, and other forces.

- I am assuming boats can't hold position without circling, so will be "like FW".

----

> Submarines (UUVs) actively brake, and hold the current horizontal position and depth, countering external forces.

- I am assuming UUVs are like MCs, but perhaps this is not true? Or perhaps different subs can do different things. 

> Submarines (UUVs) stabilise their orientation and depth. Vehicles that can actively break and hold position in the horizontal plane will do so, those that cannot will hold their current track against any external forces.

----

> Rovers (UGVs) actively break and then stop in place.

Rovers obviously don't need the semantics of this mode. My assumption is that it is useful to have a standard mode for rover movement, and that it isn't worth defining a standard "manual unstabilised" mode that rovers use in the same way other vehicles use postion controlled (i.e. to mean "safest mode").

Reply:
> I am assuming boats can't hold position without circling, so will be "like FW".

No, it's a unique behaviour in ArduPilot (https://ardupilot.org/rover/docs/loiter-mode.html#loiter-mode).  It actually isn't that simple though, as there are a lot of boat configurations with different capabilities (sail boats, steerable motors, thrusters etc).
> Rovers (UGVs) actively break and then stop in place.

I don't think either ArduPilot or PX4 actively break in Rover.
> I am assuming UUVs are like MCs, but perhaps this is not true? Or perhaps different subs can do different things.

Massively variable, and UUV vs ROV vs Glider tend to have different behaviours.  I'm not sure that boat or ROV/UUV should be included in this for now.

Reply:
Thanks. They aren't included for now. This is just exploratory, and I don't think needs to be resolved now. 

That said :-)

----

> Rovers (UGVs) stop in place (cease actively moving the vehicle).

We don't "need" active breaking to be a requirement. It is more specific. We just need to agree if this is set by a GCS the user will see "reasonable behaviour for the nature of the mode".

----

> Boats (USV) must hold either position or track (the approach depends on what is possible and appropriate for the vehicle type).

That is very general but still usefully covers the user expectation that the vehicle will do one of the two things that position control enables. We could of course be more specific and require that if position holding is supported then it is always used instead of track holding - I don't know if that helps though.